src/core/outlet_observer.ts

```diff
@@ -0,0 +1,126 @@
+import { Multimap } from "../multimap"
+import { Token, TokenListObserver, TokenListObserverDelegate } from "../mutation-observers"
+import { Context } from "./context"
+import { Controller } from "./controller"
+
+import { readInheritableStaticArrayValues } from "./inheritable_statics"
+
+export interface OutletObserverDelegate {
+  outletConnected(outlet: Controller, element: Element, name: string): void
+  outletDisconnected(element: Element, name: string): void
```

Comment:
Is it possible to retrieve the `Controller` instance for an `Element` during disconnection?

I'm afraid that the difference in signatures might be surprising and not as intuitive as if they were the same. Take, for example, how the `[name]TargetConnected(target: Element)` and `[name]TargetDisconnected(target: Element)` [signatures are the same](https://stimulus.hotwired.dev/reference/targets#connected-and-disconnected-callbacks).

Reply:
I first had a matching signature, until I realized that the `controller` param was always undefined. 

The problem is, when an outlet disconnects the controller instance is not present anymore because it's already gone. Therefore it can't be passed to the callback. One option might be to keep track of the controller instance somewhere, in order to be able to look it up again to pass it over.

Reply:
> One option might be to keep track of the controller instance somewhere, in order to be able to look it up again to pass it over. 

It might be tedious to do so, but i think parity would go a long way. Maybe a WeakMap from element to Controller instance that's set when an outlet connects?

Reply:
You are right, I will look into this!

The other thing which I was thinking about is to rename the callbacks to `[outlet]OutletAppeared` and `[outlet]OutletDisapeared`. This might sound weird at first, but the thing is, if you connect a parent controller which has outlets, its outlets elements might already be "connected" (in the sense of the DOM `.isConnected`). They might not even trigger the callback in all cases because they were already connected, similar for the disconnect.

If you connect/disconnect your parent controller, should it fire the "outlet connected callback" for all outlets which were already present at the time the parent controller connects and should it fire the "outlet disconnect callback" for all outlets which will stay present after the parent controller disconnects.

It's certainly an edge-case which I think doesn't behave 100% correct at the moment.

Reply:
>  its outlets elements might already be "connected" (in the sense of the DOM `.isConnected`)

Is that different from what happens when a connected element (with connected descendants) updates its `[data-controller]` attribute to include a new controller identifier?

```js
<div id="element-without-controller">
  <div data-a-controller-target="a">A</div>
  <div data-a-controller-target="b">B</div>
</div>

<script>
const element = document.getElementById("element-without-controller")
// controller isn't "connected", but its descendant `[data-a-controller-target]` elements are
element.setAttribute("data-controller", "a-controller")
// the controller is "connected", and fires a callback for each of its descendant `[data-a-controller-target]` elements
</script>
```

Reply:
I think this is a similar scenario, and I would probably expect the same for outlet callbacks.

It just felt weird because the callbacks suggests that the element "connected" even if those elements in reality just "became relevant" for this parent controller because it marked them as one of their outlets.

Reply:
@seanpdoyle Success! `[name]OutletConnected` and `[name]OutletDisconnected` now share the same function signature via [`35b3f76`](https://github.com/marcoroth/stimulus/commit/35b3f764c5af3fb0c2b3a35d508d135af8eb6f8e).

Reply:
I was also able to resolve the "issue" where it wouldn't fire the callbacks if you dynamically added attributes which would lead that an element becomes relevant as an outlet in [`abba8db`](https://github.com/marcoroth/stimulus/commit/abba8db05b9828895b00ddb08bba876538d1d9b5) by introducing and using a new `SelectorObserver` instead of the `TokenListObserver`.